iomem_pci.c

```diff
@@ -0,0 +1,326 @@
+	struct page *page;
+	struct pci_bus_region bus_region;
+	struct resource res;
+	phys_addr_t pa;
+
+	addr = pci_alloc_p2pmem(p->pdev, PAGE_SIZE);
+	if (!addr)
+		return -ENOMEM;
+
+	page = virt_to_page(addr);
+	if (!is_zone_device_page(page)) {
+		dev_err(&p->dev,
+			"ERROR: kernel virt_to_page does not point to a ZONE_DEVICE page!");
+		err = -EFAULT;
+		goto out;
+	}
+
+	bus_region.start = pci_p2pmem_virt_to_bus(p->pdev, addr);
+	bus_region.end = bus_region.start + PAGE_SIZE;
+
+	pcibios_bus_to_resource(p->pdev->bus, &res, &bus_region);
+
+	pa = page_to_phys(page);
+	if (pa != res.start) {
+		dev_err(&p->dev,
+			"ERROR: page_to_phys does not map to the BAR address!"
+			"  %pa[p] != %pa[p]", &pa, &res.start);
+		err = -EFAULT;
+		goto out;
+	}
+
+	pa = virt_to_phys(addr);
+	if (pa != res.start) {
+		dev_err(&p->dev,
+			"ERROR: virt_to_phys does not map to the BAR address!"
+			"  %pa[p] != %pa[p]", &pa, &res.start);
+		err = -EFAULT;
+		goto out;
+	}
+
+	if (page_to_virt(page) != addr) {
+		dev_err(&p->dev,
+			"ERROR: page_to_virt does not map to the correct address!");
+		err = -EFAULT;
+		goto out;
+	}
+
+out:
+	if (err == 0)
+		dev_info(&p->dev, "kernel page mappings seem sane.");
+
+	pci_free_p2pmem(p->pdev, addr, PAGE_SIZE);
+	return err;
+}
+
+static int iomem_test_p2p_access(struct iomem_dev *p)
+{
+	u32 *addr;
+	const u32 test_value = 0x11223344;
+	int err = 0;
+
+	addr = pci_alloc_p2pmem(p->pdev, PAGE_SIZE);
+	if (!addr)
+		return -ENOMEM;
+
+	WRITE_ONCE(addr[0], 0);
+	if (READ_ONCE(addr[0]) != 0) {
+		err = -EFAULT;
+		goto out;
+	}
+
+	WRITE_ONCE(addr[0], test_value);
+	if (READ_ONCE(addr[0]) != test_value) {
+		err = -EFAULT;
+		goto out;
+	}
+
+out:
+	if (err == 0)
+		dev_info(&p->dev, "kernel can access p2p memory.");
+	else
+		dev_err(&p->dev, "ERROR: kernel can't access p2p memory!");
+
+	pci_free_p2pmem(p->pdev, addr, PAGE_SIZE);
+	return err;
+}
+
+static int iomem_test(struct iomem_dev *p)
+{
+	int err;
+
+	err = iomem_test_page_mappings(p);
+	if (err)
+		return err;
+
+	return iomem_test_p2p_access(p);
+}
+
+static void iomem_release(struct device *dev)
+{
+	struct iomem_dev *p = to_iomem(dev);
+
+	kfree(p);
+}
+
+static struct iomem_dev *iomem_create(struct pci_dev *pdev)
+{
+	struct iomem_dev *p;
+	int err;
+
+	p = kzalloc(sizeof(*p), GFP_KERNEL);
+	if (!p)
+		return ERR_PTR(-ENOMEM);
+
+	p->pdev = pdev;
+
+	device_initialize(&p->dev);
+	p->dev.class = iomem_class;
+	p->dev.parent = &pdev->dev;
+	p->dev.release = iomem_release;
+
+	p->id = ida_simple_get(&iomem_ida, 0, 0, GFP_KERNEL);
+	if (p->id < 0) {
+		err = p->id;
+		goto out_free;
+	}
+
+	dev_set_name(&p->dev, "iomem%d", p->id);
+	p->dev.devt = MKDEV(MAJOR(iomem_devt), p->id);
+
+	cdev_init(&p->cdev, &iomem_fops);
+	p->cdev.owner = THIS_MODULE;
+
+	err = cdev_device_add(&p->cdev, &p->dev);
+	if (err)
+		goto out_ida;
+
+	dev_info(&p->dev, "registered");
+
+	iomem_test(p);
+
+	return p;
+
+out_ida:
+	ida_simple_remove(&iomem_ida, p->id);
+out_free:
+	kfree(p);
```

Comment:
Per the documentation of device_initialize(), put_device() should be used to free the device after device_initalize has been called.